### DISCUSSION TITLE
How to make cvc5 search for the shortest solution?

### DISCUSSION THREAD

**Discussion by pauljurczak** [cvc5/cvc5#9799] [Q&A] [answered]
Given this problem definition:
```
(set-logic LIA)

(synth-fun f ((x Int) (y Int)) Int
    
    ((I Int) (B Bool))

    ((I Int (x y 0 1 2 3 4 5 6 7 8 9
             (+ I I) (- I I)
             (ite B I I)))
     (B Bool ((and B B) (or B B) (not B)
              (= I I) (<= I I) (>= I I))))
)

(declare-var x Int)
(declare-var y Int)

(constraint (= (f 3 1) 3))
(constraint (= (f 2 2) 2))
(constraint (= (f 1 3) 1))
(constraint (= (f 4 4) 4))
(constraint (= (f 8 5) 9))
(constraint (= (f 7 6) 9))
(constraint (= (f 9 7) 9))
(constraint (= (f 5 8) 9))
(constraint (= (f 6 9) 9))

(check-synth)
```
cvc5 produces a somewhat convoluted solution:
```
(
(define-fun f ((x Int) (y Int)) Int (ite (= x y) x (ite (= x 1) x (ite (= x 3) x 9))))
)
```
I have to really narrow down the grammar to
```
(set-logic LIA)

(synth-fun f ((x Int) (y Int)) Int
    
    ((I Int) (B Bool))

    ((I Int (x y 1 2 3 4 5 6 7 8 9
             (ite B I I)))
     (B Bool ((<= I I))))
)

(declare-var x Int)
(declare-var y Int)

(constraint (= (f 3 1) 3))
(constraint (= (f 2 2) 2))
(constraint (= (f 1 3) 1))
(constraint (= (f 4 4) 4))
(constraint (= (f 8 5) 9))
(constraint (= (f 7 6) 9))
(constraint (= (f 9 7) 9))
(constraint (= (f 5 8) 9))
(constraint (= (f 6 9) 9))

(check-synth)
```
in order to get the shortest, I know of, solution:
```
(
(define-fun f ((x Int) (y Int)) Int (ite (<= x 4) x 9))
)
```
What settings, both for CLI and Python API use case, should I use to make cvc5 look for the shortest solution?

**Comment by ajreynol** [comment, marked as answer]
Hi,

You can use `--no-sygus-pbe`.

With this option, cvc5 produces this solution after ~4 seconds:
```
(
(define-fun f ((x Int) (y Int)) Int (ite (<= y 4) x 9))
)
```

In detail, by default cvc5 will use decision tree learning to quickly find solutions for PBE problems + ITE.  Disabling this option leads to solutions that are minimal, but notice that it may take cvc5 longer to find them.

Hope this helps,
Andy

**Reply by pauljurczak** [reply]
Great! Thanks a lot. How do I use this option with Python API? `solver.setOption("sygus-pbe", "false")`? `solver.setOption("no-sygus-pbe", "true")`?

**Comment by ajreynol** [comment]
Use `solver.setOption("sygus-pbe", "false")`. 

By convention, cvc5's command line options `X` that are Boolean have the syntax `--no-X` to specify that `X` should have value `false`.
